controllers/serviceaccount_controller.go

```diff
@@ -0,0 +1,879 @@
+		// Until Grafana 12.0.2 there was no reliable way to detect a 404 when deleting a service account.
+		// The API still returns 500 (see https://github.com/grafana/grafana/issues/106618).
+		//
+		// Once we upgrade to Grafana > 12.0.2 and bump github.com/grafana/grafana-openapi-client-go,
+		// we can handle the real 404 explicitly.
+		//
+		// In the meantime we treat any non-nil error from the delete call as "already removed" and just log it for visibility.
+		logf.FromContext(ctx).Error(err, "failed to delete service account (may already be deleted)",
+			"serviceAccountID", cr.Status.Account.ID,
+			"serviceAccountName", cr.Spec.Name,
+		)
+		// return fmt.Errorf("deleting service account %q: %w", status.SpecID, err)
+	}
+
+	return nil
+}
+
+// lookupGrafana retrieves the Grafana instance referenced by the GrafanaServiceAccount
+// and validates that it's in a ready state for accepting API requests.
+func (r *GrafanaServiceAccountReconciler) lookupGrafana(
+	ctx context.Context,
+	cr *v1beta1.GrafanaServiceAccount,
+) (*v1beta1.Grafana, error) {
+	var grafana v1beta1.Grafana
+
+	err := r.Get(ctx, client.ObjectKey{
+		Namespace: cr.Namespace,
+		Name:      cr.Spec.InstanceName,
+	}, &grafana)
+	if err != nil {
+		return nil, err
+	}
+
+	// Check if Grafana instance is ready
+	if grafana.Status.Stage != v1beta1.OperatorStageComplete || grafana.Status.StageStatus != v1beta1.OperatorStageResultSuccess {
+		return nil, fmt.Errorf("Grafana instance %q is not ready (stage: %q, status: %q)", cr.Spec.InstanceName, grafana.Status.Stage, grafana.Status.StageStatus) // nolint:staticcheck
+	}
+
+	return &grafana, nil
+}
+
+// reconcileWithInstance performs the core reconciliation logic for active resources.
+//
+// It orchestrates the complete synchronization process:
+// 1. Ensures the service account exists in Grafana (creates if missing)
+// 2. Updates service account properties to match the spec
+// 3. Manages the lifecycle of authentication tokens and their secrets
+func (r *GrafanaServiceAccountReconciler) reconcileWithInstance(
+	ctx context.Context,
+	gClient *genapi.GrafanaHTTPAPI,
+	cr *v1beta1.GrafanaServiceAccount,
+) error {
+	err := r.upsertAccount(ctx, gClient, cr)
+	if err != nil {
+		return fmt.Errorf("upserting service account: %w", err)
+	}
+
+	updateForm := buildUpdateFormIfNeeded(&cr.Spec, cr.Status.Account)
+	if updateForm != nil {
+		update, err := gClient.ServiceAccounts.UpdateServiceAccount(
+			service_accounts.
+				NewUpdateServiceAccountParamsWithContext(ctx).
+				WithServiceAccountID(cr.Status.Account.ID).
+				WithBody(updateForm),
+		)
+		if err != nil {
+			return fmt.Errorf("updating service account: %w", err)
+		}
+
+		cr.Status.Account.IsDisabled = update.Payload.Serviceaccount.IsDisabled
+		cr.Status.Account.Role = update.Payload.Serviceaccount.Role
+		cr.Status.Account.Name = update.Payload.Serviceaccount.Name
+	}
```

Comment:
I think we can simplify this by unconditionally updating the account every time

Reply:
However, this means making an extra call, even if unnecessary, on almost every reconciliation cycle (every 10 minutes) for each Grafana service account. Is it acceptable?

Reply:
HTTP request + DB write operation. :) Well, then account retrieval looks redundant. Got rid of this step too.